File: packages/indexing-module/src/publisher.ts
import type { StorageAdapter, DeliveryAdapter, AnchorAdapter, AnchorEntry } from "./adapters/types.js";
import { type DocumentEnvelope, ENVELOPE_SCHEMA, canonicalize, validateEnvelope } from "./envelope.js";
import { sha256 } from "./hash.js";
import { retry, type RetryOptions } from "./retry.js";

export interface PublisherOptions {
  storage: StorageAdapter;
  delivery: DeliveryAdapter;
  /** Optional — only required if `anchor()` is invoked on the publisher side. */
  anchor?: AnchorAdapter;
  /** Logos Delivery topic. Defaults to the canonical Whistleblower topic. */
  topic?: string;
  /** Retry policy for storage uploads. */
  retry?: RetryOptions;
  /** Clock injection point for deterministic tests. */
  now?: () => number;
}

export interface PublishResult {
  envelope: DocumentEnvelope;
  /** SHA-256 over the canonical envelope bytes; the value committed on-chain. */
  metadataHash: Uint8Array;
}

/**
 * Publisher: the "uploader-side" half of the pipeline.
 *
 * Workflow:
 *   1. `publish(bytes, meta)` — upload to Logos Storage with retry,
 *      build the canonical envelope, broadcast on Logos Delivery.
 *   2. `anchor(result)` — OPTIONAL, only if the publisher wants to anchor
 *      their own document immediately. Equivalent functionality is also
 *      offered to any third party via the `BatchAnchor` class.
 *
 * The two halves are decoupled by design: a publisher with no on-chain
 * funds can still upload+broadcast and rely on an external batch anchorer.
 */
export class Publisher {
  private readonly topic: string;
  private readonly now: () => number;

  constructor(private readonly opts: PublisherOptions) {
    this.topic = opts.topic ?? "/logos/whistleblower/v1/documents";
    this.now = opts.now ?? (() => Date.now());
  }

  async publish(
    bytes: Uint8Array,
    meta: {
      title: string;
      description: string;
      content_type: string;
      tags?: string[];
    },
  ): Promise<PublishResult> {
    const uploaded = await retry(
      () => this.opts.storage.put(bytes, { contentType: meta.content_type }),
      this.opts.retry,
    );

    const envelope: DocumentEnvelope = {
      schema: ENVELOPE_SCHEMA,
      cid: uploaded.cid,
      title: meta.title,
      description: meta.description,
      content_type: meta.content_type,
      size_bytes: uploaded.sizeBytes,
      timestamp: this.now(),
      ...(meta.tags && meta.tags.length > 0 ? { tags: meta.tags } : {}),
    };
    validateEnvelope(envelope);

    const metadataHash = await sha256(canonicalize(envelope));

    // Delivery publish: idempotent on (topic, cid) at the adapter layer.
    await this.opts.delivery.publish(this.topic, envelope);

    return { envelope, metadataHash };
  }

  /**
   * Publisher-side anchor — distinct from the basic upload+broadcast path.
   * Idempotent: if the registry already has this CID it is a no-op.
   */
  async anchor(result: PublishResult): Promise<{ tx: string; alreadyAnchored: boolean }> {
    if (!this.opts.anchor) {
      throw new Error("anchor adapter not configured");
    }
    const entry: AnchorEntry = { cid: result.envelope.cid, metadataHash: result.metadataHash };
    if (await this.opts.anchor.isAnchored(entry.cid)) {
      return { tx: "", alreadyAnchored: true };
    }
    const receipt = await this.opts.anchor.anchorSingle(entry);
    return { tx: receipt.tx, alreadyAnchored: receipt.newlyAnchored === 0 };
  }
}

File: packages/whistleblower-core/src/publisher.rs
use std::sync::Arc;
use std::time::{SystemTime, UNIX_EPOCH};

use crate::adapters::{AnchorAdapter, AnchorEntry, DeliveryAdapter, StorageAdapter};
use crate::envelope::{validate, DocumentEnvelope, ENVELOPE_SCHEMA};
use crate::hash::metadata_hash;
use crate::retry::{retry, RetryConfig};

#[derive(Clone)]
pub struct PublisherConfig {
    pub topic: String,
    pub retry: RetryConfig,
}

impl Default for PublisherConfig {
    fn default() -> Self {
        Self {
            topic: crate::DEFAULT_DELIVERY_TOPIC.into(),
            retry: RetryConfig::default(),
        }
    }
}

pub struct Publisher {
    storage: Arc<dyn StorageAdapter>,
    delivery: Arc<dyn DeliveryAdapter>,
    anchor: Option<Arc<dyn AnchorAdapter>>,
    cfg: PublisherConfig,
}

#[derive(Debug, Clone)]
pub struct PublishResult {
    pub envelope: DocumentEnvelope,
    pub metadata_hash: [u8; 32],
}

#[derive(Debug, Clone)]
pub struct PublishMeta {
    pub title: String,
    pub description: String,
    pub content_type: String,
    pub tags: Vec<String>,
}

impl Publisher {
    pub fn new(
        storage: Arc<dyn StorageAdapter>,
        delivery: Arc<dyn DeliveryAdapter>,
        anchor: Option<Arc<dyn AnchorAdapter>>,
        cfg: PublisherConfig,
    ) -> Self {
        Self {
            storage,
            delivery,
            anchor,
            cfg,
        }
    }

    pub async fn publish(&self, bytes: &[u8], meta: PublishMeta) -> anyhow::Result<PublishResult> {
        let storage = self.storage.clone();
        let bytes_owned = bytes.to_vec();
        let ct = meta.content_type.clone();
        let upload = retry(
            self.cfg.retry,
            // Treat anyhow errors as transient by default. Production code can
            // wrap a typed error and discriminate on `is_transient` more
            // precisely (e.g. don't retry 4xx).
            |_e: &anyhow::Error| true,
            move || {
                let storage = storage.clone();
                let bytes = bytes_owned.clone();
                let ct = ct.clone();
                async move { storage.put(&bytes, Some(&ct)).await }
            },
        )
        .await
        .map_err(|e| anyhow::anyhow!("upload failed: {}", e))?;

        let envelope = DocumentEnvelope {
            schema: ENVELOPE_SCHEMA.into(),
            cid: upload.cid,
            title: meta.title,
            description: meta.description,
            content_type: meta.content_type,
            size_bytes: upload.size_bytes,
            timestamp: now_ms(),
            tags: if meta.tags.is_empty() {
                None
            } else {
                Some(meta.tags)
            },
        };
        validate(&envelope)?;

        let hash = metadata_hash(&envelope)?;
        self.delivery.publish(&self.cfg.topic, &envelope).await?;

        Ok(PublishResult {
            envelope,
            metadata_hash: hash,
        })
    }

    /// Publisher-side anchor — the optional "anchor on-chain" action that's
    /// distinct from the basic upload flow.
    pub async fn anchor(&self, result: &PublishResult) -> anyhow::Result<AnchorOutcome> {
        let anchor = self
            .anchor
            .as_ref()
            .ok_or_else(|| anyhow::anyhow!("anchor adapter not configured"))?;
        if anchor.is_anchored(&result.envelope.cid).await? {
            return Ok(AnchorOutcome {
                tx: String::new(),
                already_anchored: true,
            });
        }
        let receipt = anchor
            .anchor_single(AnchorEntry {
                cid: result.envelope.cid.clone(),
                metadata_hash: result.metadata_hash,
            })
            .await?;
        Ok(AnchorOutcome {
            tx: receipt.tx,
            already_anchored: receipt.newly_anchored == 0,
        })
    }
}

#[derive(Debug, Clone)]
pub struct AnchorOutcome {
    pub tx: String,
    pub already_anchored: bool,
}

fn now_ms() -> u64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_millis() as u64)
        .unwrap_or(0)
}
